hw/xbox/xid-sdl.c

```diff
@@ -194,36 +198,49 @@ static const XIDDesc desc_xid_xbox_gamepad = {
 
 #define BUTTON_MASK(button) (1 << ((button) - GAMEPAD_DPAD_UP))
 
-static void update_output(USBXIDState *s) {
-#ifdef FORCE_FEEDBACK
-    if (s->sdl_haptic == NULL) { return; }
+static int haptic_effect_prepare(SDL_HapticEffect *effect,
+                              HapticEmulationMode mode,
+                              uint16_t large_magnitude,
+                              uint16_t small_magnitude)
+{
+    switch (mode) {
+    case EMU_HAPTIC_LEFT_RIGHT:
+        effect->leftright = (SDL_HapticLeftRight){
+            .type = SDL_HAPTIC_LEFTRIGHT,
+            .length = SDL_HAPTIC_INFINITY,
+            /* Xbox rumble returns in a range of 0x0000 -> 0xFFFF,
+             * SDL requires the range to be within 0x0000 -> 0x7FFF
+             */
+            .large_magnitude = large_magnitude >> 1,
```

Comment:
After testing this PR, I don't think these `.large_magnitude` and `.small_magnitude` fields actually max out at `0x7FFF` as we had previously thought. Testing the haptic effects myself revealed that using `0x7FFF` has the rumble motors running well below full power. When I set the magnitude to `0xFFFF` however, the motors run at (what *feels like*) full power--certainly a noticeable difference over `0x7FFF`. So, I think we should probably pass the values here as-is and remove the comment/shift.

Reply:
You are probably running an old version. This upstream bug was only fixed in May 2018, so the fix might not have landed yet in all distros:

https://github.com/spurious/SDL-mirror/commit/6066a1cf5fb6862f15cedaf6b3a81826203f78a1

We should implement features as documented and report issues upstream. Eventually all distros will be fixed.